File: 1301-1400/1361-Validate Binary Tree Nodes/1361-Validate Binary Tree Nodes.py
class Solution:
    def validateBinaryTreeNodes(self, n: int, leftChild: List[int], rightChild: List[int]) -> bool:
        degree = [0] * n
        for left, right in zip(leftChild, rightChild):
            if left != -1:
                degree[left] += 1
                if degree[left] > 1:
                    return False
            if right != -1:
                degree[right] += 1
                if degree[right] > 1:
                    return False
                
        root = -1
        for i, d in enumerate(degree):
            if d == 0:
                if root == -1:
                    root = i
                else:
                    return False
        
        def dfs(curr):
            if curr == -1:
                return 0
            return 1 + dfs(leftChild[curr]) + dfs(rightChild[curr])
        
        return dfs(root) == n

File: 1301-1400/1361-Validate Binary Tree Nodes/1361-Validate Binary Tree Nodes.cpp
class Solution {
public:
    bool validateBinaryTreeNodes(int n, vector<int>& leftChild, vector<int>& rightChild) {
        vector<int> degree(n);
        for (int i = 0; i < n; ++i) {
            if (leftChild[i] != -1 && ++degree[leftChild[i]] > 1) {
                return false;
            }
            if (rightChild[i] != -1 && ++degree[rightChild[i]] > 1) {
                return false;
            }
        }
        
        int root = -1;
        for (int i = 0; i < n; ++i) {
            if (degree[i] == 0) {
                if (root == -1) {
                    root = i;
                }
                else {
                    return false;
                }
            }
        }
        
        return dfs(root, leftChild, rightChild) == n;
    }
    
private:
    int dfs(int curr, vector<int>& leftChild, vector<int>& rightChild) {
        if (curr == -1) {
            return 0;
        }
        return 1 + dfs(leftChild[curr], leftChild, rightChild) + dfs(rightChild[curr], leftChild, rightChild);
    }
};
